### PR TITLE
Add countLeadingZeros

### PR DESCRIPTION
**Pull request by evahop**
Adds `countLeadingZeros` from https://github.com/gfx-rs/naga/issues/2134#issue-1463112793

For negative expressions return `0`, otherwise return `31 - msb`.

This assumes `findUMsb`, `findMSB`, and `firstbithigh` all return an index from the `lsb`,
but I don't have the means to verify this is actually the case atm.

### CODE REVIEW

**Review by teoxoy — CHANGES_REQUESTED** [review]
Behavior looks good!
But the GLSL and HLSL backends write the expr multiple times which won't always work.

The expression needs to either be baked or the function needs to be polyfilled.
I'd go with baking for now since we have the mechanism already implemented for GLSL.

https://github.com/gfx-rs/naga/blob/c7d02151f08d6285683795289b5725b827d836d1/src/back/glsl/mod.rs#L1767-L1768

I feel like polyfilling would be a nicer solution but we'd ideally make the transition all at once.
See also https://github.com/gfx-rs/naga/issues/2123 for more details.

**Comment by evahop** [comment]
> The expression needs to either be baked or the function needs to be polyfilled. I'd go with baking for now since we have the mechanism already implemented for GLSL.

Okay, I see the issue.
I'm just now noticing `FindMSB` also has a minimum requirement of `400` while the `README.md` pegs it at `330+`.
Would it be preferable to replace it with something along the lines of [es.math.clz32.js](https://github.com/zloirock/core-js/blob/5385428316f590f4845006e2b15e24db3be4d7e9/packages/core-js/modules/es.math.clz32.js)?

**Comment by evahop** [comment]
I meant `uints` in the last commit but I assume it's going to get squashed either way.
Otherwise that should do it, it should address the issue in https://github.com/gfx-rs/naga/issues/2123#issue-1441029855,
and I can follow up with a PR if the `findMSB`/`findLSB` need replacing.

**Comment by evahop** [comment]
Had some downtime so just went ahead and did it for the scope of this PR and won’t likely have time to address it later.

**Comment by evahop** [comment]
I see what you’re pointing out. You can rebase and open another issue. I tried applying the same polyfill  to `findMSB` just now, using the `((clz + 1) % 33) …` method on the `mls` backend, and I’m not sure if its producing the expected behavior, I need to mess with it some more.

File: src/back/glsl/mod.rs [hunk 1]
```diff
@@ -2941,33 +2932,55 @@ impl<'a, W: Write> Writer<'a, W> {
...
                                 if let crate::ScalarKind::Uint = kind {
-                                    write!(self.out, "uvec{s}(ivec{s}(31) - findMSB(")?;
+                                    write!(self.out, "uvec{s}(")?;
+                                } else {
+                                    write!(self.out, "ivec{s}(")?;
+                                }
+
+                                write!(self.out, "vec{s}(31.0) - floor(log(vec{s}(")?;
+                                self.write_expr(arg, ctx)?;
+                                write!(self.out, ") + 0.5) * {LOG2_E})), ")?;
```
Comment (teoxoy) [inline src/back/glsl/mod.rs]: Can we use `log2(x)` instead of `log(x) * LOG2E`?

**Comment by teoxoy** [comment]
> If I’m not mistaken, seems like `clz` will always produce a `0` for negative numbers, while `findMSB` returns an index of the first set `0`.

This sounds correct, so the msl polyfill is not correct for negative numbers.

File: src/back/glsl/mod.rs [hunk 2]
```diff
@@ -2930,6 +2931,56 @@ impl<'a, W: Write> Writer<'a, W> {
...
+                                    self.write_expr(arg, ctx)?;
+                                    write!(self.out, ", ivec{s}(0))))")?;
+                                }
+                            }
+                            crate::TypeInner::Scalar { kind, .. } => {
+                                write!(self.out, "(")?;
+                                self.write_expr(arg, ctx)?;
+
+                                if let crate::ScalarKind::Uint = kind {
+                                    write!(self.out, " == 0u ? 32u : uint(")?;
```
Comment (teoxoy) [inline src/back/glsl/mod.rs:2968]: For `uint`, I don't think we need an extra `if`. `floor(log2(0.5)) = -1`

File: src/back/glsl/mod.rs (hunk 1 above)
Comment (evahop) [reply src/back/glsl/mod.rs]: Yes, feel free to make whatever changes. I just transcribed it verbatim at the time

**Pull request merged**